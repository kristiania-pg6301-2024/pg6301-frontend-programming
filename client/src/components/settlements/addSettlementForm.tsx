import React from "react";

export function AddSettlementForm() {
  return (
    <form>
      <h1>Add settlement</h1>
      <button>Submit</button>
    </form>
  );
}
