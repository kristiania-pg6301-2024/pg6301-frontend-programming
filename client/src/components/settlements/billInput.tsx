import { BillType } from "./domain";
import React from "react";

interface Props {
  type: BillType;
  value?: number | undefined;
  setValue: (v: number) => void;
}

export function BillInput({ type: { key }, value, setValue }: Props) {
  return (
    <div key={key}>
      <label htmlFor={`newSettlementBalance_${key}`}>{key}:</label>
      <input
        id={`newSettlementBalance_${key}`}
        type={"number"}
        value={value || "0"}
        onChange={(e) => setValue(parseInt(e.target.value))}
      />
    </div>
  );
}
