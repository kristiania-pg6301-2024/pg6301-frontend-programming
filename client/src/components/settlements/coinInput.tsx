import { CoinAmount, CoinType } from "../../../../src/cashBalance";
import React, { Dispatch, SetStateAction } from "react";

interface Props {
  type: CoinType;
  value?: CoinAmount;
  setValue: Dispatch<SetStateAction<CoinAmount>>;
}

export function CoinInput({
  type: { key, gramsPerCoin },
  value,
  setValue,
}: Props) {
  return (
    <div key={key} className={`denomination${key}`}>
      <label>
        {key}:
        <input
          type={"number"}
          className={"count"}
          disabled={value && "grams" in value}
          value={
            !value
              ? ""
              : "count" in value
                ? value.count
                : Math.round(value.grams / gramsPerCoin)
          }
          onChange={(e) => setValue({ count: parseInt(e.target.value) })}
        />
        <input
          type={"number"}
          className={"grams"}
          disabled={value && "count" in value}
          value={
            !value
              ? ""
              : "grams" in value
                ? value.grams
                : value.count * gramsPerCoin
          }
          onChange={(e) => setValue({ grams: parseFloat(e.target.value) })}
        />
      </label>
    </div>
  );
}
