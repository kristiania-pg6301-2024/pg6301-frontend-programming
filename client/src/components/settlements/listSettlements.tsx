import { Settlement, sumBalance } from "../../../../src/cashBalance";
import { Link } from "react-router-dom";
import React from "react";

interface Props {
  settlements: Settlement[];
}

export function ListSettlements({ settlements }: Props) {
  return (
    <>
      <h1>Settlements</h1>
      {settlements.map(({ _id, department, balance }) => (
        <div key={_id}>
          {department}: {sumBalance(balance)} kr
        </div>
      ))}
      <Link to={"/new"}>New Settlement</Link>
    </>
  );
}
