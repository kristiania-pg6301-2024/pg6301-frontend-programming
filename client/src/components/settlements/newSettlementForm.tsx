import React, { useState } from "react";
import { billTypes, CashBalance, coinTypes } from "../../../../src/cashBalance";

const departments = ["Furniture", "Books"];

export interface Settlement {
  department: string;
  balance: CashBalance;
}

interface Props {
  onNewSettlement(settlement: Settlement): void;
}

export function NewSettlementForm({ onNewSettlement }: Props) {
  const [department, setDepartment] = useState("");
  const [balance, setBalance] = useState<CashBalance>({});
  function handleSubmit(e: React.FormEvent) {
    e.preventDefault();
    onNewSettlement({ department, balance });
  }

  function hasBalance() {
    return Object.values(balance).some((b) => !!b);
  }

  return (
    <form onSubmit={handleSubmit}>
      <h1>New Settlement</h1>

      <div>
        <label htmlFor={"newSettlementDepartment"}>Department: </label>
        <select
          id={"newSettlementDepartment"}
          value={department}
          onChange={(e) => setDepartment(e.target.value)}
        >
          <option></option>
          {departments.map((department) => (
            <option key={department}>{department}</option>
          ))}
        </select>
      </div>

      {billTypes.map(({ key }) => (
        <div key={key}>
          <label htmlFor={`newSettlementBalance_${key}`}>{key}:</label>
          <input
            id={`newSettlementBalance_${key}`}
            type={"number"}
            value={balance[key] || "0"}
            onChange={(e) =>
              setBalance((old) => ({ ...old, [key]: parseInt(e.target.value) }))
            }
          />
        </div>
      ))}
      <div className={"coins"}>
        {coinTypes.map(({ key, gramsPerCoin }) => (
          <div key={key} className={`denomination${key}`}>
            <label>
              {key}:
              <input
                type={"number"}
                className={"count"}
                disabled={balance[key] && "grams" in balance[key]}
                value={
                  !balance[key]
                    ? ""
                    : "count" in balance[key]
                      ? balance[key].count
                      : Math.round(balance[key].grams / gramsPerCoin)
                }
                onChange={(e) =>
                  setBalance((old) => ({
                    ...old,
                    [key]: { count: parseInt(e.target.value) },
                  }))
                }
              />
              <input
                type={"number"}
                className={"grams"}
                disabled={balance[key] && "count" in balance[key]}
                value={
                  !balance[key]
                    ? ""
                    : "grams" in balance[key]
                      ? balance[key].grams
                      : balance[key].count * gramsPerCoin
                }
                onChange={(e) =>
                  setBalance((old) => ({
                    ...old,
                    [key]: { grams: parseFloat(e.target.value) },
                  }))
                }
              />
            </label>
          </div>
        ))}
      </div>
      <button type={"submit"} disabled={!department || !hasBalance()}>
        Submit
      </button>
    </form>
  );
}
