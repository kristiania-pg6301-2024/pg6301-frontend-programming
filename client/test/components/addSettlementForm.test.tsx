import React from "react";

import { expect, test } from "vitest";
import { render } from "@testing-library/react";
import { AddSettlementForm } from "../../src/components/settlements/addSettlementForm";

test("settlement form", () => {
  const app = render(<AddSettlementForm />);
  expect(app.baseElement).toMatchSnapshot();
});
