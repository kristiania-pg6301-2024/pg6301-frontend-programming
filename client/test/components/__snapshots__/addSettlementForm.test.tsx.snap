// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`settlement form 1`] = `
<body>
  <div>
    <form>
      <h1>
        Add settlement
      </h1>
      <button>
        Submit
      </button>
    </form>
  </div>
</body>
`;
